torch/csrc/jit/serialization/export_module.cpp

```diff
@@ -415,18 +426,44 @@ void ScriptModuleSerializer::writeArchive(
       [&](const c10::ClassTypePtr& t) {
         return type_name_uniquer_.getUniqueName(t);
       },
-      &memoizedClassTypes);
+      &memoizedClassTypes,
+      [&](const at::Tensor& tensor) {
+        // returns a string to use in picker.cpp as storage obj key
+        if (tensor_cdata_naming_scheme) {
+          tensor_names.push_back(
+              std::to_string(reinterpret_cast<std::intptr_t>(
+                  tensor.storage().unsafeGetStorageImpl())) +
+              ".storage");
+        } else {
+          tensor_names.push_back(std::to_string(tensor_names.size()));
+        }
+        return tensor_names.back();
+      });
   data_pickle.protocol();
   data_pickle.pushIValue(value);
   data_pickle.stop();
+  // write out tensor data
   size_t i = 0;
-  std::string prefix = archive_name + "/";
+  TORCH_INTERNAL_ASSERT(tensor_names.size() == data_pickle.tensorData().size());
+  const std::vector<std::string>& pre_serialized_files =
+      writer_.getAllWrittenRecords();
+
   for (const auto& td : data_pickle.tensorData()) {
     WriteableTensorData writable_td = getWriteableTensorData(td);
-    std::string fname = prefix + c10::to_string(i++);
-    writer_.writeRecord(fname, writable_td.data(), writable_td.sizeInBytes());
+    std::string file_name = tensor_dir + tensor_names[i++];
+    if (tensor_cdata_naming_scheme &&
+        std::find(
```

Comment:
maybe build an unordered_set so that we don't have to do a linear search for every tensor?

Reply:
I was thinking that the tradeoff of copying the vector into an unordered set would be more costly than just doing the linear search. Right now the `writer_.getAllWrittenRecords()` returns a reference to its contents and isn't doing any copying or duplication (I think that's how cpp works?). I'm not sure if it's more expensive to do a lot of string comparisons or if it's more expensive to copy all of the string up front to do less comparisons